src/Microsoft.AspNet.Diagnostics/ErrorPageMiddleware.cs

```diff
@@ -182,31 +188,103 @@ private StackFrame StackFrame(Chunk line)
 
             int lineNumber = line.ToInt32();
 
-            return string.IsNullOrEmpty(file)
-                ? LoadFrame(string.IsNullOrEmpty(function) ? line.ToString() : function, string.Empty, 0)
-                : LoadFrame(function, file, lineNumber);
+            if (string.IsNullOrEmpty(file))
+            {
+                return GetStackFrame(
+                    string.IsNullOrEmpty(function) ? line.ToString() : function,
+                    file: string.Empty,
+                    lineNumber: 0,
+                    fileProvider: _options.FileProvider,
+                    sourceCodeLineCount: _options.SourceCodeLineCount);
+            }
+            else
+            {
+                return GetStackFrame(function, file, lineNumber, _options.FileProvider, _options.SourceCodeLineCount);
+            }
         }
 
-        private StackFrame LoadFrame(string function, string file, int lineNumber)
+        internal static StackFrame GetStackFrame(
+            string function,
+            string file,
+            int lineNumber,
+            IFileProvider fileProvider,
+            int sourceCodeLineCount)
         {
             var frame = new StackFrame { Function = function, File = file, Line = lineNumber };
+
+            if (string.IsNullOrEmpty(file))
+            {
+                return frame;
+            }
+
+            IEnumerable<string> lines = null;
             if (File.Exists(file))
             {
-                var code = File.ReadLines(file);
-                ReadFrameContent(frame, code, lineNumber, lineNumber);
+                lines = File.ReadLines(file);
             }
+            else
+            {
+                // Handle relative paths and embedded files
+                var fileInfo = fileProvider.GetFileInfo(file);
+                if (fileInfo.Exists)
+                {
+                    // ReadLines doesn't accept a stream. Use ReadLines as its more efficient
+                    // relative to reading lines via stream reader
+                    if (!string.IsNullOrEmpty(fileInfo.PhysicalPath))
+                    {
+                        lines = File.ReadLines(fileInfo.PhysicalPath);
+                    }
+                    else
+                    {
+                        lines = GetLines(fileInfo.CreateReadStream());
+                    }
+                }
+            }
+
+            if (lines != null)
+            {
+                GetErrorCodeBlock(frame, lines, lineNumber, sourceCodeLineCount);
+            }
+
             return frame;
         }
 
-        private void ReadFrameContent(StackFrame frame,
-                                      IEnumerable<string> code,
-                                      int startLineNumber,
-                                      int endLineNumber)
+        private static void GetErrorCodeBlock(
```

Comment:
`GetErrorCodeBlock` => `ReadErrorCodeBlock` or `SetErrorCodeBlock` (or something better if you can think of one). `Get` doesn't really make sense here since it's a void returning method.